### PR TITLE
Dashboard: Panel edit and support for different layout items

### PR DESCRIPTION
**Pull request by torkelo**
Fixes bug introduced in https://github.com/grafana/grafana/pull/95768  (changed the useMemo in away so that repeat options are not working as it relied on unused memo dependency). 

This PR changes how options are rendered so we do not need re-build the full options list for options to re-render or for option visibility to be re-evaluated. Now options can control this themselves and we no longer need to rely on passing unused objects to useMemo dependency array. 

* DashboardLayoutItem interface 
* Use layout item interface in panel edit when rendering options
* Make it possible to have hooks in panel edit options functions (to subscribe to state changes)
* Make it possible to control panel edit option visibility via hooks

### CODE REVIEW

File: public/app/features/dashboard-scene/scene/DashboardPanelMenu.tsx
Comment (oscarkilhed) [inline public/app/features/dashboard-scene/scene/DashboardPanelMenu.tsx:1]: This pretty much replaces `public/app/features/dashboard-scene/scene/PanelMenuBehavior.tsx` right? Should we remove these functions from that file?

Comment (torkelo) [reply public/app/features/dashboard-scene/scene/DashboardPanelMenu.tsx:1]: Sorry this file is from a different file poc, must have been unsaved in vs code and saved accidentally, will remove it

File: public/app/features/dashboard-scene/panel-edit/getPanelFrameOptions.tsx [hunk 1]
```diff
@@ -30,7 +25,7 @@ export function getPanelFrameCategory2(
   const panelLinksObject = dashboardSceneGraph.getPanelLinks(panel);
   const links = panelLinksObject?.state.rawLinks ?? [];
   const dashboard = getDashboardSceneFor(panel);
-  const layoutElement = panel.parent;
+  const layoutElement = panel.parent!;
```
Comment (dprokop) [inline public/app/features/dashboard-scene/panel-edit/getPanelFrameOptions.tsx:28]: nit: a missing rename to `layoutItem`

File: public/app/features/dashboard-scene/panel-edit/getPanelFrameOptions.tsx [hunk 2]
```diff
@@ -97,72 +92,8 @@ export function getPanelFrameCategory2(
...
-              onChange={(value) => gridItem.setState({ maxPerRow: value.value })}
-            />
-          );
-        },
-      })
-    );
-
-    descriptor.addCategory(category);
+  if (isDashboardLayoutItem(layoutElement) && layoutElement.getOptions) {
+    descriptor.addCategory(layoutElement.getOptions());
```
Comment (dprokop) [inline public/app/features/dashboard-scene/panel-edit/getPanelFrameOptions.tsx:96]: Nice, this is great abstraction.

**Review by oscarkilhed — APPROVED** [review]

File: public/app/features/dashboard/components/PanelEditor/OptionsPaneItemDescriptor.tsx [hunk 3]
```diff
@@ -29,74 +31,79 @@ export interface OptionsPaneItemProps {
...
 
-    return (
-      <Label description={description && this.highlightWord(description, searchQuery)} category={categories}>
-        {this.highlightWord(title, searchQuery)}
-        {overrides && overrides.length > 0 && <OptionsPaneItemOverrides overrides={overrides} />}
-      </Label>
-    );
+  // eslint-disable-next-line react-hooks/rules-of-hooks
+  if (useShowIf && !useShowIf()) {
+    return null;
```
Comment (dprokop) [inline public/app/features/dashboard/components/PanelEditor/OptionsPaneItemDescriptor.tsx:52]: How about making useShowIf an instance method on `OptionsPaneItemDescriptor` that would return either `false` by default or configured hook? Then we wouldn't have to depend on reading descriptor's props and executing hooks conditionaly.

Comment (torkelo) [reply public/app/features/dashboard/components/PanelEditor/OptionsPaneItemDescriptor.tsx]: @dprokop good idea! moved this to a useShowIf function on OptionsPaneItemDescriptor the nice thing with that is that that function can also check the old showif function and use that if that is defined, that way we only need one if statement in the render function

**Pull request merged**